### PR TITLE
Added Subscription and Notification Endpoints

### PR DESCRIPTION
**Pull request by mubashir494**
Tasks Done
- Added Endpoints to subscribe and unsubscribe
- Added Enpoints to get Notification
- Added Endpoints to get count of Notification
- Integerated Pusher for realtime Communication
- Added Unit test for Subscribe Endpoints

Closes : #20

### CODE REVIEW

File: Backend/Backend.Tests/SubscriptionTest.cs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            }
+            catch (Exception e)
+            {
+                Assert.Fail(e.Message);
+
+            }
+            Assert.True(response);
+        }
+
+        //[Fact]
```
Comment (BigSaladMan) [inline Backend/Backend.Tests/SubscriptionTest.cs:147]: Test is commented out

File: Backend/Backend.Tests/SubscriptionTest.cs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+                {
+                    HttpContext = _httpContextMock.Object
+                }
+            };
+
+            MockBasicUtilities(out var podcast, out var user, out var episode);
+        }
+
+        [Fact]
+        public void Subscribe_Podcast_Success()
```
Comment (BigSaladMan) [inline Backend/Backend.Tests/SubscriptionTest.cs:90]: Make your test titles a little more specific. Should be something like this
Subscription_.SubscribeAsync_ValidRequest_Success 
in other words:
What feature you are testing, what function you are testing, what you are testing about it, and what the expected result is

Comment (BigSaladMan) [reply Backend/Backend.Tests/SubscriptionTest.cs:90]: makes it easier to search for this, and when it fails its quicker to notice what is wrong

File: Backend/Backend/Controllers/NotificationController.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+            // If User is not found, return 404
+            if (user is null)
+                return NotFound("User does not exist.");
+
+            return Ok(new { count =  await _notificationService.GetUnreadNoticationCountAsync(user) });
+        }
+
+
+
```
Comment (BigSaladMan) [inline Backend/Backend/Controllers/NotificationController.cs:54]: Not sure where to put this, but you forgot to add tests for this. I can do this if you want tho

**Review by BigSaladMan — APPROVED** [review]
looks pretty good, if you want I will add the remaining tests in this and we can merge it

**Comment by BigSaladMan** [comment]
Will need to look into why the end-to-end tests are broken as well

**Comment by BigSaladMan** [comment]
Do not Close this convo, waiting for tests to be added to the two controllers before merging

File: Backend/Backend.Tests/SubscriptionTest.cs (hunk 1 above)
Comment (BigSaladMan) [reply Backend/Backend.Tests/SubscriptionTest.cs:147]: Dont close one of these convos, cus I need to add the tests still and closing the convos will let the merge happen

File: Backend/Backend/Infrastructure/AppDbContext.cs [hunk 4]
```diff
@@ -146,7 +146,7 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             .HasMany(e => e.Ratings)
             .WithOne(e => e.Podcast)
             .HasForeignKey(e => e.PodcastId)
-            .IsRequired();
+            .OnDelete(DeleteBehavior.Restrict);
```
Comment (BigSaladMan) [inline Backend/Backend/Infrastructure/AppDbContext.cs:149]: Fairly certain you do not want this restricted. If you delete a podcast, the ratings should come with it

Comment (mubashir494) [reply Backend/Backend/Infrastructure/AppDbContext.cs:149]: Actually, there were two 1 to many relation on rating entity so it was causing error

File: Backend/Backend/Infrastructure/AppDbContext.cs [hunk 5]
```diff
@@ -33,6 +32,7 @@ public AppDbContext(string connectionString) : base(new DbContextOptionsBuilder<
     public virtual DbSet<PlaylistElement> PlaylistElements { get; set; }
     public virtual DbSet<Playlist> Playlists { get; set; }
     public virtual DbSet<Files>? File { get; set; }
+    public virtual DbSet<Notification>? Notifications { get; set; }
```
Comment (mathieutrudeau) [inline Backend/Backend/Infrastructure/AppDbContext.cs:35]: Avoid making the dB sets nullable.

**Review by mathieutrudeau — APPROVED** [review]

File: Backend/Backend/Controllers/NotificationController.cs (hunk 3 above)
Comment (BigSaladMan) [reply Backend/Backend/Controllers/NotificationController.cs]: redid all tests. Should be fine

File: Backend/Backend.Tests/SubscriptionTest.cs (hunk 2 above)
Comment (BigSaladMan) [reply Backend/Backend.Tests/SubscriptionTest.cs]: redid all tests. Should be fine

File: Backend/Backend.Tests/SubscriptionTest.cs (hunk 1 above)
Comment (BigSaladMan) [reply Backend/Backend.Tests/SubscriptionTest.cs]: redid all tests. Should be fine

**Pull request merged**